drivers/hmc5883l/hmc5883l.c

```diff
@@ -0,0 +1,301 @@
+        DEBUG("[hmc5883l] %s i2c dev=%d addr=%02x: " f "\n", \
+              __func__, d->dev, HMC5883L_I2C_ADDRESS, ## __VA_ARGS__);
+
+#else /* ENABLE_DEBUG */
+
+#define DEBUG_DEV(f, d, ...)
+
+#endif /* ENABLE_DEBUG */
+
+#define ERROR_DEV(f, d, ...) \
+        LOG_ERROR("[hmc5883l] %s i2c dev=%d addr=%02x: " f "\n", \
+                  __func__, d->dev, HMC5883L_I2C_ADDRESS, ## __VA_ARGS__);
+
+#define EXEC_RET(f, r) \
+    if ((r = f) != HMC5883L_OK) { \
+        DEBUG("[hmc5883l] %s: error code %d\n", __func__, res); \
+        return res; \
+    }
+
+#define EXEC_RET_CODE(f, r, c) \
+    if ((r = f) != HMC5883L_OK) { \
+        DEBUG("[hmc5883l] %s: error code %d\n", __func__, res); \
+        return c; \
+    }
+
+/** Forward declaration of functions for internal use */
+
+static int _is_available(const hmc5883l_t *dev);
+
+static int _reg_read(const hmc5883l_t *dev, uint8_t reg, uint8_t *data, uint16_t len);
+static int _reg_write(const hmc5883l_t *dev, uint8_t reg, uint8_t data);
+
+int hmc5883l_init(hmc5883l_t *dev, const hmc5883l_params_t *params)
+{
+    int res = HMC5883L_OK;
+
+    assert(dev != NULL);
+    assert(params != NULL);
+    DEBUG_DEV("params=%p", dev, params);
+
+    /* init sensor data structure */
+    dev->dev = params->dev;
+#if MODULE_HMC5883L_INT
+    dev->int_pin = params->int_pin;
+#endif
+    dev->op_mode = params->op_mode;
+    dev->gain = params->gain;
+
+    /* check availability of the sensor */
+    EXEC_RET(_is_available(dev), res)
+
+    /* set configuration register A and B */
+    EXEC_RET(_reg_write(dev, HMC5883L_REG_CFG_A, params->meas_avg |
+                                                 params->meas_mode |
+                                                 params->dor), res);
+    EXEC_RET(_reg_write(dev, HMC5883L_REG_CFG_B, params->gain), res);
+
+    /* set operation mode */
+    EXEC_RET(_reg_write(dev, HMC5883L_REG_MODE, params->op_mode), res);
+
+    /* to set the LOCK flag, read the first data sample that is not valid */
+    uint8_t data[6];
+    EXEC_RET(_reg_read(dev, HMC5883L_REG_OUT_X_MSB, data, 6), res);
+
+    return res;
+}
+
+#ifdef MODULE_HMC5883L_INT
+
+int hmc5883l_init_int(hmc5883l_t *dev, hmc5883l_drdy_int_cb_t cb, void *arg)
+{
+    assert(dev != NULL);
+    assert(dev->int_pin != GPIO_UNDEF);
+    DEBUG_DEV("", dev);
+
+    if (gpio_init_int(dev->int_pin, GPIO_IN, GPIO_FALLING, cb, arg)) {
+        return HMC5883L_ERROR_COMMON;
+    }
+
+    return HMC5883L_OK;
+}
+
+#endif /* MODULE_HMC5883L_INT */
+
+int hmc5883l_data_ready(const hmc5883l_t *dev)
+{
+    assert(dev != NULL);
+    DEBUG_DEV("", dev);
+
+    int res = HMC5883L_OK;
+
+    uint8_t reg;
+
+    EXEC_RET(_reg_read(dev, HMC5883L_REG_STATUS, &reg, 1), res);
+    return (reg == HMC5883L_REG_STATUS_RDY) ? HMC5883L_OK : HMC5883L_ERROR_NO_DATA;
+}
+
+/*
+ * Scale factors for conversion of raw sensor data to uGs for possible
+ * sensitivities according to the datasheet.
+ */
+static const int32_t HMC5883L_RES[] = {
```

Comment:
```suggestion
static const int16_t HMC5883L_RES[] = {
```

should be enough here.